### PR TITLE
Add initial shaping performance test and profiler

### PR DESCRIPTION
**Pull request by ebraminio**
My personal code related to #566 inspired from https://github.com/harfbuzz/harfbuzz/pull/1398#discussion_r235566351 and https://youtu.be/nXaxk27zwlk

### CODE REVIEW

**Comment by ebraminio** [comment]
image

**Comment by ebraminio** [comment]
Arabic text with Amiri
image

**Comment by behdad** [comment]
Can we reuse hb-shape?  It has all parameters for it.  This is how I use it:
```
$ time perf stat ./hb-shape -O '' -o /dev/null Roboto-Regular.ttf --text-file thelittleprince.txt  --num-iterations=100 --font-funcs=ot
```

The only reason to write C code for perf is to integrate with a perf library, which I like us to do.  I remember Google had a library for this.  Looks like https://github.com/gperftools/gperftools ?  Or not?

**Comment by behdad** [comment]
Oh, maybe https://github.com/google/benchmark

**Comment by behdad** [comment]
I also have had good luck w sysprof recently.

**Comment by behdad** [comment]
As for test matrix:

* Long paragraphs.  For Latin thelittleprince is good.  For Persian, in Pango we have a looong single-paragraph monologue.  Use that.  I have permission for it.  https://github.com/GNOME/pango/blob/master/utils/test-long-paragraph.txt

* Words.  To emulate systems that do word-caching.  The Wikipedia word-list for various languages from https://github.com/harfbuzz/harfbuzz-testing-wikipedia or for ASCII, `/usr/share/dict/words`.

As for fonts, we care about testing at least:

* Roboto.
* Noto Nastaliq Urdu.
* Amiri.
* Noto Sans Devanagari.

**Comment by ebraminio** [comment]
> Can we reuse hb-shape?

> Oh, maybe https://github.com/google/benchmark

You should want either this or that and the current is also neither of them :) (wanted to go for bench, first am experiencing if can I for simd first)

**Comment by behdad** [comment]
Yeah I want to try SIMD as well if I find a free afternoon... Just timing hb-shape with the right parameters should be fine for that one.

**Comment by behdad** [comment]
Or, not time, but `perf stat`.  On an otherwise-idle laptop, the instruction-count I get from `perf stat` is super deterministic.  Also pay attention to `insn per cycle`.

**Comment by behdad** [comment]
Can't we use hb-shape instead of adding new source for that?  I use:
```
perf stat ./hb-shape -O '' -o /dev/null NotoNastaliq-Regular.ttf --text-file test-long-paragraph.txt  --num-iterations=100  --font-funcs=ot
```

**Comment by ebraminio** [comment]
Then guess all I want is to have one linear command to compile hb-shape, let's close, will post accordingly

**Comment by ebraminio** [comment]
`clang++ -DPACKAGE_NAME='""' -DPACKAGE_VERSION='""' util/hb-shape.cc util/options.cc src/harfbuzz.cc -Isrc -DHAVE_GLIB $(pkg-config --cflags --libs glib-2.0) -DHB_NO_MT -fno-exceptions -fno-rtti -o hb-shape` does the trick

**Comment by ebraminio** [comment]
Actually the rest can be useful, let's update this with the command

**Comment by ebraminio** [comment]
Have a look Behdad

**Comment by behdad** [comment]
Thanks.  We're thinking along the same lines indeed.  I'll comment with my wishlist here later today.

**Comment by behdad** [comment]
Rough notes / suggestions:

- Add toplevel perf/ dir indeed.

- Add test-long-paragraph.txt and thelittheprince.txt, for Arabic and Latin scripts.  Can rename them to something more useful.  Should check how long paragraphs in thelittleprince.txt are.  Maybe also add a wordlist, to emulate word-cache workflows.

- Add Roboto (whatever latest version Android ships), NotoNastaliqUrdu, Amiri, and maybe NotoSansDevanagari.  Find good Devanagari text to add above as well.

- Add `--perf` to `hb-shape` that sets the defaults.  Ie: `./hb-shape -O '' -o /dev/null --num-iterations=100 --font-funcs=ot`

- Add a perf framework ala what we have in our test framework, that implements measuring time elapsed, repeating test, dropping outliers, and summarizing numbers.  This is to be used to write micro-benchmarks, like reporting bsearch vs ksearch performance for various number of input elements.  I really need this test!  Same for measuring set operations performance and possible vectorization speedup there again.

- By default no special optimization flags (unroll loops, etc) should be specified.  We should be measuring 1. defaults (which should have -O2 already), 2. -O2, 3. -O3 with `-march=native -mtune=native` perhaps.

**Comment by behdad** [comment]
We can then define a perf suite that is simply pairs of font-file and text-file.  Later we will try running and collecting those numbers online on variety of machines over time perhaps maybe.

File: test/perf/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+hb-shape
```
Comment (behdad) [inline test/perf/.gitignore]: We don't keep .gitignore in tree.

Comment (ebraminio) [reply test/perf/.gitignore]: Done

**Comment by ebraminio** [comment]
> Add toplevel perf/ dir indeed.

Done

> Add test-long-paragraph.txt and thelittheprince.txt, for Arabic and Latin scripts. Can rename them to something more useful. Should check how long paragraphs in thelittleprince.txt are. Maybe also add a wordlist, to emulate word-cache workflows.

Already done except wordlist, is this https://github.com/sindresorhus/word-list/blob/master/words.txt ok? Is 2mb…

> Add Roboto (whatever latest version Android ships), NotoNastaliqUrdu, Amiri, and maybe NotoSansDevanagari. Find good Devanagari text to add above as well.

Done

> Add --perf to hb-shape that sets the defaults. Ie: ./hb-shape -O '' -o /dev/null --num-iterations=100 --font-funcs=ot
> Add a perf framework ala what we have in our test framework, that implements measuring time elapsed, repeating test, dropping outliers, and summarizing numbers. This is to be used to write micro-benchmarks, like reporting bsearch vs ksearch performance for various number of input elements. I really need this test! Same for measuring set operations performance and possible vectorization speedup there again.

Maybe later? I mean, let's have something to hack upon it over the time, think that should work also.

Current output is:

```
 Performance counter stats for './hb-shape -o /dev/null fonts/NotoNastaliqUrdu-Regular.ttf --text-file text/test-long-paragraph.txt --num-iterations=100 --font-funcs=ot':

          5,003.70 msec task-clock:u              #    1.000 CPUs utilized          
                 0      context-switches:u        #    0.000 K/sec                  
                 0      cpu-migrations:u          #    0.000 K/sec                  
               541      page-faults:u             #    0.108 K/sec                  
    17,819,606,588      cycles:u                  #    3.561 GHz                    
    50,330,220,517      instructions:u            #    2.82  insn per cycle         
     8,486,482,328      branches:u                # 1696.041 M/sec                  
        58,516,636      branch-misses:u           #    0.69% of all branches        

       5.005475650 seconds time elapsed

       4.983649000 seconds user
       0.000000000 seconds sys
````

> By default no special optimization flags (unroll loops, etc) should be specified. We should be measuring 1. defaults (which should have -O2 already), 2. -O2, 3. -O3 with -march=native -mtune=native perhaps.

Done

**Comment by behdad** [comment]
Looks good for starters.  I like to add the `--perf` to hb-shape as well.  Not a huge fan of `run` script.  Prefer using existing build system.  But yeah, happy to merge this.  Specially since it doesn't go into the tarball yet.  Feel free to merge.  We change from there.

Let's rename `test-long-paragraph.txt` to something else as well, reflecting what's inside maybe.

**Comment by ebraminio** [comment]
> Let's rename test-long-paragraph.txt to something else as well, reflecting what's inside maybe.

Done

**Pull request merged**